io/jvm/src/main/scala/fs2/io/net/SocketPlatform.scala

```diff
@@ -112,23 +112,47 @@ private[net] trait SocketCompanionPlatform {
       writeSemaphore: Semaphore[F]
   )(implicit F: Async[F])
       extends BufferedReads[F](readSemaphore) {
+
+    @volatile private[this] var inputOpen = true
+    @volatile private[this] var outputOpen = true
+
     protected def readChunk(buffer: ByteBuffer): F[Int] =
-      F.async_[Int] { cb =>
+      F.async[Int] { cb =>
         ch.read(
           buffer,
           null,
-          new IntCallbackHandler(cb)
+          new CompletionHandler[Integer, AnyRef] {
+            def completed(bytesRead: Integer, attachment: AnyRef) =
+              cb(Right(bytesRead))
+            def failed(err: Throwable, attachment: AnyRef) =
+              if (inputOpen) cb(Left(err))
+          }
         )
+        F.delay(Some(F.delay {
+          inputOpen = false
+          ch.shutdownInput()
```

Comment:
Sorry, I got confused why we need these open flags. By the time we make it to this block where we do `ch.shutdownInput()` the `cb` function is already a no-op right?

Reply:
Fixed...